### PR TITLE
enabled moving of graph channels

### PR DESCRIPTION
**Pull request by MartinKT**

### CODE REVIEW

**Review by fvacek — CHANGES_REQUESTED** [review]

File: libshvvisu/src/timeline/graphwidget.h [hunk 1]
```diff
@@ -62,14 +64,23 @@ class SHVVISU_DECL_EXPORT GraphWidget : public QWidget
...
 	QSize m_graphPreferredSize;
-	enum class MouseOperation { None = 0, MiniMapLeftResize, MiniMapRightResize, ChannelHeaderResize, MiniMapScrollZoom, GraphDataAreaLeftPress, GraphDataAreaLeftCtrlPress, GraphDataAreaLeftCtrlShiftPress, GraphAreaMove, GraphAreaSelection, GraphDataAreaRightPress};
+	enum class MouseOperation { None = 0, MiniMapLeftResize, MiniMapRightResize, ChannelHeaderResize, MiniMapScrollZoom, ChannelHeaderMove, GraphDataAreaLeftPress, GraphDataAreaLeftCtrlPress, GraphDataAreaLeftCtrlShiftPress, GraphAreaMove, GraphAreaSelection, GraphDataAreaRightPress };
 	MouseOperation m_mouseOperation = MouseOperation::None;
 	QPoint m_recentMousePos;
 	int m_resizeChannelIx = -1;
+	QPoint m_channelScreenShotOffset;
+	QLabel *m_channelScreenShot = nullptr;
+	QTimer m_mouseMoveScrollTimer;
+	QWidget *m_thickYellowLine;
```
[the remark this answers is not in the document]
Comment (MartinKT) [reply libshvvisu/src/timeline/graphwidget.h:82]: Opraveno.

File: libshvvisu/src/timeline/graphwidget.h [hunk 2]
```diff
@@ -62,14 +64,23 @@ class SHVVISU_DECL_EXPORT GraphWidget : public QWidget
 	bool isMouseAboveMiniMapSlider(const QPoint &pos) const;
 	int posToChannelVerticalHeader(const QPoint &pos) const;
 	int posToChannel(const QPoint &pos) const;
+	void scrollOnMouseMove();
+	bool scrollOnMouseMove(const QPoint &mouse_pos);
```
[the remark this answers is not in the document]
Comment (MartinKT) [reply libshvvisu/src/timeline/graphwidget.h:68]: Presne tak.

[the remark this answers is not in the document]
Comment (fvacek) [reply libshvvisu/src/timeline/graphwidget.h:73]: pak by se mela jmenovat `scrollByMouseOuterOverlap`

File: libshvvisu/src/timeline/graphwidget.h [hunk 3]
```diff
@@ -62,14 +64,23 @@ class SHVVISU_DECL_EXPORT GraphWidget : public QWidget
 	bool isMouseAboveMiniMapSlider(const QPoint &pos) const;
 	int posToChannelVerticalHeader(const QPoint &pos) const;
 	int posToChannel(const QPoint &pos) const;
+	void scrollOnMouseMove();
```
[the remark this answers is not in the document]
Comment (fvacek) [reply libshvvisu/src/timeline/graphwidget.h]: ???

**Comment by fvacek** [comment]
co ty ostatni commenty?

File: libshvvisu/src/timeline/graphwidget.cpp [hunk 4]
```diff
@@ -636,6 +770,39 @@ void GraphWidget::contextMenuEvent(QContextMenuEvent *event)
...
+	QPoint pos = event->pos();
+
+	if (scrollOnMouseMove(mapToGlobal(pos))) {
+		m_channelHeaderMoveContext->mouseMoveScrollTimer->start();
+	}
+	else {
+		m_channelHeaderMoveContext->mouseMoveScrollTimer->stop();
+	}
+
+	moveYellowThickLineAccordingToPos(pos);
```
Comment (fvacek) [inline libshvvisu/src/timeline/graphwidget.cpp:794]: `moveDropMarker...`

File: libshvvisu/src/timeline/graphwidget.cpp [hunk 5]
```diff
@@ -456,6 +466,54 @@ void GraphWidget::mouseMoveEvent(QMouseEvent *event)
 
 		return;
 	}
+	case MouseOperation::ChannelHeaderMove:
+		m_mouseOperation = MouseOperation::None;
+		if (gr->channelCount() == 0) {
+			return;
+		}
+
+			QRect header_rect;
```
Comment (fvacek) [inline libshvvisu/src/timeline/graphwidget.cpp:475]: indentation

File: libshvvisu/src/timeline/graphwidget.cpp [hunk 6]
```diff
@@ -456,6 +466,54 @@ void GraphWidget::mouseMoveEvent(QMouseEvent *event)
...
+			QPalette pal = m_channelHeaderMoveContext->channelDropMarker->palette();
+			pal.setColor(QPalette::ColorRole::Window, Qt::yellow);
+			m_channelHeaderMoveContext->channelDropMarker->setAutoFillBackground(true);
+			m_channelHeaderMoveContext->channelDropMarker->setPalette(pal);
+			m_channelHeaderMoveContext->channelDropMarker->resize(header_rect.width(), QFontMetrics(font()).height() / 2);
+			m_channelHeaderMoveContext->channelDropMarker->show();
+			drag->exec();
+			event->accept();
+			setAcceptDrops(false);
+			delete m_channelHeaderMoveContext->mouseMoveScrollTimer;
```
Comment (fvacek) [inline libshvvisu/src/timeline/graphwidget.cpp:510]: should be in m_channelHeaderMoveContext destructor

File: libshvvisu/src/timeline/graphwidget.cpp [hunk 7]
```diff
@@ -456,6 +466,54 @@ void GraphWidget::mouseMoveEvent(QMouseEvent *event)
...
+			delete m_channelHeaderMoveContext->channelDropMarker;
```
Comment (fvacek) [inline libshvvisu/src/timeline/graphwidget.cpp:511]: should be in m_channelHeaderMoveContext destructor

File: libshvvisu/src/timeline/graphwidget.cpp [hunk 8]
```diff
@@ -456,6 +466,54 @@ void GraphWidget::mouseMoveEvent(QMouseEvent *event)
...
+			delete m_channelHeaderMoveContext;
```
Comment (fvacek) [inline libshvvisu/src/timeline/graphwidget.cpp:512]: I think that this is possible memory leak to have delete only in this if branch

**Review by fvacek — CHANGES_REQUESTED** [review]

**Pull request merged**